dev/src/document-reader.ts

```diff
@@ -0,0 +1,193 @@
+        orderedDocuments.push(finalDoc.build());
+      } else {
+        throw new Error(`Did not receive document for "${docRef.path}".`);
+      }
+    }
+
+    return orderedDocuments;
+  }
+
+  private async fetchAllDocuments(requestTag: string): Promise<void> {
+    while (this.remainingDocuments.size > 0) {
+      try {
+        return await this.fetchMoreDocuments(requestTag);
+      } catch (err) {
+        // If a non-transactional read failed, attempt to restart.
+        // Transactional reads are retried via the transaction runner.
+        if (
+          this.transactionId ||
+          isPermanentRpcError(err, 'batchGetDocuments')
+        ) {
+          logger(
+            'DocumentReader.fetchAllDocuments',
+            requestTag,
+            'BatchGetDocuments failed with non-retryable stream error:',
+            err
+          );
+          throw err;
+        } else {
+          logger(
+            'DocumentReader.fetchAllDocuments',
+            requestTag,
+            'BatchGetDocuments failed with retryable stream error:',
+            err
+          );
+        }
+      }
+    }
+  }
+
+  private fetchMoreDocuments(requestTag: string): Promise<void> {
+    const request: api.IBatchGetDocumentsRequest = {
+      database: this.firestore.formattedName,
+      transaction: this.transactionId,
+      documents: Array.from(this.remainingDocuments),
+    };
+
+    if (this.fieldMask) {
+      const fieldPaths = this.fieldMask.map(
+        fieldPath => (fieldPath as FieldPath).formattedName
+      );
+      request.mask = {fieldPaths};
+    }
+
+    let resultCount = 0;
+
+    return this.firestore
+      .requestStream('batchGetDocuments', request, requestTag)
+      .then(stream => {
+        return new Promise<void>((resolve, reject) => {
+          stream
+            .on('error', err => reject(err))
+            .on('data', (response: api.IBatchGetDocumentsResponse) => {
+              try {
+                let document;
+
+                if (response.found) {
+                  logger(
+                    'DocumentReader.fetchMoreDocuments',
+                    requestTag,
+                    'Received document: %s',
+                    response.found.name!
+                  );
+                  document = this.firestore.snapshot_(
+                    response.found,
+                    response.readTime!
+                  );
+                } else {
+                  logger(
+                    'DocumentReader.fetchMoreDocuments',
+                    requestTag,
+                    'Document missing: %s',
+                    response.missing!
+                  );
+                  document = this.firestore.snapshot_(
+                    response.missing!,
+                    response.readTime!
+                  );
+                }
+
+                const path = document.ref.formattedName;
+                this.remainingDocuments.delete(path);
```

Comment:
optional: add assert() to check that path is in the set.

Reply:
Skipped for now.